tavern/_plugins/mqtt/response.py

```diff
@@ -181,11 +181,14 @@ def addwarning(w, *args, **kwargs):
 
             self._maybe_run_validate_functions(msg)
         else:
-            self._adderr(
-                "Expected '%s' on topic '%s' but no such message received",
-                expected_payload,
-                topic,
-            )
+            if self.expected.get("unexpected"):
+                return {}
+            else:
+                self._adderr(
+                    "Expected '%s' on topic '%s' but no such message received",
+                    expected_payload,
+                    topic,
+                )
```

Comment:
```suggestion
            if not self.expected.get("unexpected"):
                self._adderr(
                    "Expected '%s' on topic '%s' but no such message received",
                    expected_payload,
                    topic,
                )
```

Returning early will break the saving of variables from responses, this should do what is required

Reply:
Indeed: [tavern.txt](https://github.com/taverntesting/tavern/files/10071620/tavern.txt)

Reply:
My initial submission works, your suggestion including my extra attempts to fix it, doesn't. Any other ideas/suggestions?

Reply:
Sorry, I've been looking at the 'multiple mqtt responses' feature in Tavern 2.0 and wasn't thinking about there only being one mqtt response.

I think that what needs to be done is that my suggestion does need to be applied, but then also just after it checks if there were any errors (https://github.com/taverntesting/tavern/blob/540588f1a50adaccc1e16bd4032ba55d12233da5/tavern/_plugins/mqtt/response.py#L190-L197) it should do 
```python
if not msg:
    return {}
```
because there won't be any variables to save.

Reply:
Made a slight alteration to your suggestion, seems to work like that. Let me know if that's OK for you?